### DISCUSSION TITLE
m-form: @validation-success and @validation-error are not triggered

### DISCUSSION THREAD

**Discussion by iliubinskii** [quasarframework/quasar#12711] [General - Components / Directives / etc] [answered]
### What happened?

@validation-error is not triggered though validation fails and error message shows.
@validation-success is also not triggered.

### What did you expect to happen?

My goal is to enable reset button only when something changes, i.e. when either:
1. User input changes or
2. Validation state switches to invalid

I expect to receive @validation-error whenever form becomes invalid and @validation-success whenever it becomes valid.
This way I can track form validation state.

### Reproduction URL

https://codepen.io/ilyalyu/pen/eYexwZy?editors=1111

### How to reproduce?

1. Open https://codepen.io/ilyalyu/pen/eYexwZy?editors=1111
2. Open console
3. Enter input
4. Leave input without entering anything
5. You will see error message under input field
6. But console is empty

### Flavour

Quasar CLI (@quasar/cli | @quasar/app)

### Areas

Components (quasar)

**Comment by MilosPaunovic** [comment]
As per docs, `@validation-success` and `@validation-error` events are emitted after a validation was triggered.

**Comment by iliubinskii** [comment]
> As per docs, @validation-success and @validation-error events are emitted after a validation was triggered.

Ok, but in this case how can I track validation state in my codepen example?
It seems there is no way.

**Comment by frankmichel** [comment]
@MilosPaunovic I am running into the same issue. Is there any way to check the validation status without actually triggering the validation in the ux?

`$refs.form.isValid()`

would be an awesome addition to `.validate()` on the q-form.

**Reply by MilosPaunovic** [reply]
At the moment, no, you have to validate it manually.

**Comment by joaofranciscoguarda** [comment]
Same problem here, trying to make a button not be clickable until inputs are valid, but it seems @validation-success and @validation-error not work. 
Would be nice if docs say that it may not work, but ok, validate manually isnt a big deal.

**Reply by metalsadman** [reply]
If you put your form inputs inside a QForm it has a method that has ability to get all form components inside it, then you can make a computed value or watcher that check each of the components `hasError` computed prop, then have a flag, let's say `isDirty` that you can then check on your q-btn's/q-form's submit/disable props. Ideally make a composable out of it. I made one myself but slightly different as I made wrappers for the quasar input components.

![image](https://github.com/quasarframework/quasar/assets/31038838/3c59ba1e-04bc-4fd9-8c49-78817824f303)

**Comment by metalsadman** [comment, marked as answer]
I have something like this 
```ts
// intersection type of known Quasar form components
export type FormValidationComponent = QInput | QSelect | QField;

/**
 * adds form validation aware for form components
 * @formRef = template ref of QForm container
 * @formComponents = ref that will hold array of form components
 * @formIsDirty = ref sets state of a form component if its value was changed
 * @formHasError = computed ref sets if one of the form component has error
 */
export function useFormValidation() {
  // assign this in the consumer's QForm template ref
  const formRef = ref<QForm>();
  const formComponents = ref<
    {
      uid: string;
      validationComponent: FormValidationComponent;
      origModel: unknown;
    }[]
  >();
  const formIsDirty = computed(() =>
    formComponents.value
      ? formComponents.value.some(
          (v) => v.origModel !== v.validationComponent.modelValue
        )
      : false
  );

  // Quasar form components has internal value called hasError
  // check if the components is in error state
  const formHasError = computed(
    () =>
      formComponents.value?.some(
        (v) => v.validationComponent.hasError
      ) as boolean
  );

  async function formBindValidation() {
    // wait for vue nextTick before binding in case data are being updated
    await nextTick();
    if (formRef.value === undefined)
      throw new Error(
        'Failed to bind validation as formRef is not assigned to a QForm in the current page!'
      );
    // get all form components that has internal validations
    formComponents.value = (
      formRef.value.getValidationComponents() as FormValidationComponent[]
    ).map((v) => ({
      uid: uid(),
      validationComponent: v,
      origModel: v.modelValue,
    }));
  }

  function formReset() {
    formRef.value?.resetValidation();
    // rebind form to refresh original values if it was updated
    formBindValidation();
  }

  // watchEffect(() => {
  //   console.log('Form is dirty? ', formIsDirty.value);
  //   console.log('Form has error? ', formHasError.value);
  // });

  return {
    /**
     * must use it in your QForm to make use of validation status
     */
    formRef,
    /**
     * see form dirty status
     */
    formIsDirty,
    /**
     * see form error status
     */
    formHasError,
    /**
     * bind validations to form, must call it in your component's onMounted hook
     */
    formBindValidation,
    /**
     * resets form and re-bind validation
     */
    formReset,
  };
}
```

Some vue component
```ts
// script setup
// import useFormValidation
const { formRef, formBindValidation, formIsDirty, formHasError, formReset } =
  useFormValidation();

// bind onMounted hook
onMounted(() => {
  formBindValidation();
})
```
in the template
```html
<q-form ref="formRef" ...>
  <!-- Quasar form components inside here ... -->
  <!-- check for formHasError or formIsDirty to disable/enable button state or etc... -->
  <q-btn :disable="formHasError" .. />
</q-form>
```

**Reply by joaofranciscoguarda** [reply]
Nice job man! As you said in the other comment, a way to do it is manually, that's what I did, in these situation was just one Btn, but the next form will be huge and I will try your solution! Your code should be in Quasar, that a great hook. Thanks!

**Reply by davija** [reply]
Yes, THANK YOU!  Something like this should be built-in, that's a lot of code for something that most other frameworks provide OOTB.

**Reply by davija** [reply]
I made a slight alteration to the formHasError computed property.  I'm sure there may be a better way to do it, but the issue I ran into an issue.

Basically in the case of a fresh clean form with required fields, the current `formHasError` property will be false indicating the form is valid (even though it really isn't).

Here is my adaptation - please feel free to provide suggestions on improvement as I'm still fairly green with Vue.  Essentially if the field doesn't have an error but is pristine it will force a validation on the control and then reset the validation if there is an error and return true.

With this change, the button that is using `formHasError` will be disabled when the form first appears and will then only be enabled when the form is truly valid.  The resetting of validation happens to avoid triggering the error notification in the UI on a component that hasn't  yet been touched.

```ts  
  const formHasError = computed(
    () =>
      formComponents.value?.some(
        (v) => {
          // If field has an error, just return true now
          if (v.validationComponent.hasError) {
            return true;
          }

          // Since the field doesn't have an error, force validation
          // if the field is in a pristine state. 
          if (v.origModel === v.validationComponent.modelValue) {
            v.validationComponent.validate();

            if (v.validationComponent.hasError) {
              // If field is in an error state, reset validation 
              // After resetting validation, return true to indicate field has an error
              v.validationComponent.resetValidation();

              return true;
            }
          }

          return false;
        },
      ) as boolean,
  );
```

**Comment by fjbardelli** [comment]
Hello, does this work in quaser 2 with vue and nuxt 3?

**Reply by yusufkandemir** [reply]
See https://github.com/quasarframework/quasar/discussions/12711#discussioncomment-6050446
That solution will work with Quasar v2. Nuxt is irrelevant as it's purely related to UI logic, so it will still work.
